pyface/data_view/data_models/data_accessors.py

```diff
@@ -0,0 +1,168 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+""" Classes for extracting data from objects
+
+This module provides helper classes for the row table data model and
+related classes for extracting data from an object in a consistent way.
+"""
+
+from abc import abstractmethod
+from collections import Hashable
+
+from traits.api import (
+    ABCHasStrictTraits, Any, Event, Instance, Int, Str, observe
+)
+from traits.trait_base import xgetattr, xsetattr
+
+from pyface.data_view.abstract_data_model import DataViewSetError
+from pyface.data_view.abstract_value_type import AbstractValueType
+from pyface.data_view.value_types.api import TextValue
+
+
+class AbstractDataAccessor(ABCHasStrictTraits):
+    """ Accessor that gets and sets data on an object.
+    """
+
+    #: A human-readable label for the accessor.
+    title = Str()
+
+    #: The value type of the title of this accessor, suitable for use in a
+    #: header.
+    title_type = Instance(AbstractValueType, factory=TextValue)
+
+    #: The value type of the data accessed.
+    value_type = Instance(AbstractValueType)
+
+    #: An event fired when accessor is updated update.  The payload is
+    #: a tuple of the accessor info and whether the title or value changed
+    #: (or both).
+    updated = Event
+
+    @abstractmethod
+    def get_value(self, obj):
+        raise NotImplementedError()
+
+    def can_set_value(self, obj):
+        return False
+
+    def set_value(self, obj, value):
+        raise DataViewSetError(
+            "Cannot set {!r} column of {!r}.".format(self.title, obj)
+        )
+
+    # trait observers
+
+    @observe('title,title_type.updated')
+    def title_updated(self, event):
+        self.updated = (self, 'title')
+
+    @observe('value_type.updated')
+    def value_type_updated(self, event):
+        self.updated = (self, 'value')
+
+
+class AttributeDataAccessor(AbstractDataAccessor):
+    """ DataAccessor that presents an extended attribute on an object.
+
+    This is suitable for use with Python objects, including HasTraits
+    classes.
+    """
+
+    #: The extended attribute name of the trait holding the value.
+    attr = Str()
+
+    def get_value(self, obj):
+        return xgetattr(obj, self.attr, None)
```

Comment:
I will think about this.  The current expectation of the API is that there will be no errors returned by get methods, because early on I ran into issues with performance using exception handling higher up the stack.  This may have been due to relative likelihoods of exceptions.

But I think I'd like to defer that investigation to a separate issue.

I will look into some sort of logged message or warning if there is a missing attribute, so at least it isn't silenced.

Reply:
On customizing the default value:

- the `value_type` can be customized to return something different (particularly for the text value), and that's a better place for it, I think.
- people can always write their own subclasses

So I'm not too concerned about that right now.

Reply:
> I will think about this. The current expectation of the API is that there will be no errors returned by get methods, because early on I ran into issues with performance using exception handling higher up the stack. This may have been due to relative likelihoods of exceptions.
>
> But I think I'd like to defer that investigation to a separate issue.

Yep. Let's open a separate issue then. It's probably a good idea to consider this across all the objects we have.

> I will look into some sort of logged message or warning if there is a missing attribute, so at least it isn't silenced.

Yes please.